// react-history/BrowserHistory.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createBrowserHistory = require('history/createBrowserHistory');

var _createBrowserHistory2 = _interopRequireDefault(_createBrowserHistory);

var _PropTypes = require('./PropTypes');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Manages session history using the HTML5 history API including
 * pushState, replaceState, and the popstate event.
 */
var BrowserHistory = function (_React$Component) {
  _inherits(BrowserHistory, _React$Component);

  function BrowserHistory() {
    _classCallCheck(this, BrowserHistory);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  BrowserHistory.prototype.getChildContext = function getChildContext() {
    return { history: this.history };
  };

  BrowserHistory.prototype.componentWillMount = function componentWillMount() {
    var _this2 = this;

    var _props = this.props,
        basename = _props.basename,
        forceRefresh = _props.forceRefresh,
        getUserConfirmation = _props.getUserConfirmation,
        keyLength = _props.keyLength;


    this.history = (0, _createBrowserHistory2.default)({
      basename: basename,
      forceRefresh: forceRefresh,
      getUserConfirmation: getUserConfirmation,
      keyLength: keyLength
    });

    // Do this here so we catch actions in cDM.
    this.unlisten = this.history.listen(function () {
      return _this2.forceUpdate();
    });
  };

  BrowserHistory.prototype.componentWillUnmount = function componentWillUnmount() {
    this.unlisten();
  };

  BrowserHistory.prototype.render = function render() {
    var children = this.props.children;


    return typeof children === 'function' ? children(this.history) : _react2.default.Children.only(children);
  };

  return BrowserHistory;
}(_react2.default.Component);

BrowserHistory.propTypes = {
  basename: _react.PropTypes.string,
  forceRefresh: _react.PropTypes.bool,
  getUserConfirmation: _react.PropTypes.func,
  keyLength: _react.PropTypes.number,
  children: _react.PropTypes.oneOfType([_react.PropTypes.node, _react.PropTypes.func]).isRequired
};
BrowserHistory.childContextTypes = {
  history: _PropTypes.history.isRequired
};
exports.default = BrowserHistory;
