// react-history/MemoryHistory.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createMemoryHistory = require('history/createMemoryHistory');

var _createMemoryHistory2 = _interopRequireDefault(_createMemoryHistory);

var _PropTypes = require('./PropTypes');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Manages session history using in-memory storage.
 */
var MemoryHistory = function (_React$Component) {
  _inherits(MemoryHistory, _React$Component);

  function MemoryHistory() {
    _classCallCheck(this, MemoryHistory);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  MemoryHistory.prototype.getChildContext = function getChildContext() {
    return { history: this.history };
  };

  MemoryHistory.prototype.componentWillMount = function componentWillMount() {
    var _this2 = this;

    var _props = this.props,
        getUserConfirmation = _props.getUserConfirmation,
        initialEntries = _props.initialEntries,
        initialIndex = _props.initialIndex,
        keyLength = _props.keyLength;


    this.history = (0, _createMemoryHistory2.default)({
      getUserConfirmation: getUserConfirmation,
      initialEntries: initialEntries,
      initialIndex: initialIndex,
      keyLength: keyLength
    });

    // Do this here so we catch actions in cDM.
    this.unlisten = this.history.listen(function () {
      return _this2.forceUpdate();
    });
  };

  MemoryHistory.prototype.componentWillUnmount = function componentWillUnmount() {
    this.unlisten();
  };

  MemoryHistory.prototype.render = function render() {
    var children = this.props.children;


    return typeof children === 'function' ? children(this.history) : _react2.default.Children.only(children);
  };

  return MemoryHistory;
}(_react2.default.Component);

MemoryHistory.propTypes = {
  getUserConfirmation: _react.PropTypes.func,
  initialEntries: _react.PropTypes.array,
  initialIndex: _react.PropTypes.number,
  keyLength: _react.PropTypes.number,
  children: _react.PropTypes.oneOfType([_react.PropTypes.node, _react.PropTypes.func]).isRequired
};
MemoryHistory.childContextTypes = {
  history: _PropTypes.history.isRequired
};
exports.default = MemoryHistory;
